src/lib/operators/join_index.cpp

```diff
@@ -83,133 +83,124 @@ std::shared_ptr<const Table> JoinIndex::_on_execute() {
+
+          // Scan all chunks from the probe side input
+          for (ChunkID probe_chunk_id = ChunkID{0}; probe_chunk_id < _probe_input_table->chunk_count();
+               ++probe_chunk_id) {
+            const auto& probe_segment = _probe_input_table->get_chunk(probe_chunk_id)
+                                            ->get_segment(_adjusted_primary_predicate.column_ids.first);
+            segment_with_iterators(*probe_segment, [&](auto probe_iter, const auto probe_end) {
+              _reference_join_two_segments_using_index(probe_iter, probe_end, probe_chunk_id, index_chunk_id, index,
+                                                       reference_segment_pos_list);
+            });
+          }
+          performance_data.chunks_scanned_with_index++;
+        } else {
+          _fallback_nested_loop(index_chunk_id, track_probe_matches, track_index_matches, is_semi_or_anti_join,
+                                secondary_predicate_evaluator);
         }
+      } else {
+        _fallback_nested_loop(index_chunk_id, track_probe_matches, track_index_matches, is_semi_or_anti_join,
+                              secondary_predicate_evaluator);
       }
     }
-  }
-
-  // For Full Outer and Right Join we need to add all unmatched rows for the index side.
-  if (_mode == JoinMode::FullOuter || _mode == JoinMode::Right) {
-    for (ChunkID chunk_id{0}; chunk_id < _index_matches.size(); ++chunk_id) {
-      for (ChunkOffset chunk_offset{0}; chunk_offset < _index_matches[chunk_id].size(); ++chunk_offset) {
-        if (!_index_matches[chunk_id][chunk_offset]) {
-          _index_pos_list->emplace_back(RowID{chunk_id, chunk_offset});
-          _probe_pos_list->emplace_back(NULL_ROW_ID);
+  } else {  // DATA JOIN or OUTER REFERENCE JOIN
+    // Scan all chunks for index input
+    for (ChunkID index_chunk_id = ChunkID{0}; index_chunk_id < _index_input_table->chunk_count(); ++index_chunk_id) {
+      const auto& index_chunk = _index_input_table->get_chunk(index_chunk_id);
+      const auto& indexes =
+          index_chunk->get_indexes(std::vector<ColumnID>{_adjusted_primary_predicate.column_ids.second});
+
+      if (!indexes.empty()) {
+        // We assume the first index to be efficient for our join
+        // as we do not want to spend time on evaluating the best index inside of this join loop
+        const auto& index = indexes.front();
+
+        // Scan all chunks from the probe side input
+        for (ChunkID probe_chunk_id = ChunkID{0}; probe_chunk_id < _probe_input_table->chunk_count();
+             ++probe_chunk_id) {
+          const auto& probe_segment =
+              _probe_input_table->get_chunk(probe_chunk_id)->get_segment(_adjusted_primary_predicate.column_ids.first);
+          segment_with_iterators(*probe_segment, [&](auto probe_iter, const auto probe_end) {
+            _data_join_two_segments_using_index(probe_iter, probe_end, probe_chunk_id, index_chunk_id, index);
```

Comment:
I don't get this part. The `else` block handles "DATA JOIN or OUTER REFERENCE JOIN", but here you unconditionally call `_data_join_two_segments_using_index`.

Reply:
Unconditionally is not exactly true: We only do the `_data_join_two_seg...` if `!indexes.empty()`. This is only the case for chunks which contain data segments. Therefore, a chunk which contains only ref. segments cannot contain indexes because of the nature of create_index.

Please correct me if I'm wrong.

Effectively, if this branch is entered and the index_input_table is a reference table, the nested loop fallback solution is executed. This will be changes if we remove the ref. table & non-inner join support.

Reply:
Maybe add `// DATA JOIN` after checking for `!indexes.empty()`?

Reply:
Even this wouldn't be exactly true: it is also a data join if the fallback solution is used. There are currently two ways of creating indexes: On the table level (add indexes to each chunk) or on the chunk level. If the latter is used, there could be chunks in a data table without an index for a specific column, even if the previous chunk had one.

Reply:
[comment change](https://github.com/hyrise/hyrise/pull/1743/commits/72ffd8693d9a346fe84ff5ff61a1a5fd10ac73a4)